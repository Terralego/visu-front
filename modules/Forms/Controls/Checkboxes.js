import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from '@blueprintjs/core';

import './index.scss';

export class Checkboxes extends React.Component {
  static propTypes = {
    label: PropTypes.string,
    values: PropTypes.arrayOf(PropTypes.string),
    onChange: PropTypes.func,
  }

  static defaultProps = {
    label: '',
    values: [],
    onChange () {},
  }

  state = {
    values: [],
  }

  handleChange = value => () => {
    const { onChange } = this.props;
    this.setState(({ values: prevValues }) => {
      const values = prevValues.includes(value)
        ? [...prevValues.filter(val => val !== value)]
        : [...prevValues, value];
      onChange(values);
      return value;
    });
  }

  render () {
    const { label, values } = this.props;
    const { handleChange } = this;

    return (
      <div className="control-container">
        <p className="control-label">{label}</p>
        {values.map(val =>
          <Checkbox key={val} onChange={handleChange(val)} label={val} />)}
      </div>
    );
  }
}

export default Checkboxes;
