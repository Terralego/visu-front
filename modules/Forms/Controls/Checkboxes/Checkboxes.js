import React from 'react';
import PropTypes from 'prop-types';

import Checkbox from './Checkbox';
import formatValues from '../formatValues';
import '../index.scss';

export class Checkboxes extends React.Component {
  static getDerivedStateFromProps ({ values }) {
    if (!values) return null;

    return {
      values: formatValues(values),
    };
  }

  onToggle = toggledValue => {
    const { value, onChange } = this.props;
    const newValue = value.includes(toggledValue)
      ? [...value.filter(val => val !== toggledValue)]
      : [...value, toggledValue];
    onChange(newValue);
  }

  render () {
    const { value: mainValue, label: mainLabel } = this.props;
    const { values } = this.state;
    const { onToggle } = this;

    return (
      <div className="control-container">
        <p className="control-label">{mainLabel}</p>
        {values.map(({ label, value }) => (
          <Checkbox
            key={`${label}${value}`}
            onToggle={onToggle}
            value={value}
            label={label}
            checked={mainValue.includes(value)}
          />
        ))}
      </div>
    );
  }
}


Checkboxes.propTypes = {
  label: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
  // Prop is used in getDerivedStateFromProps
  // eslint-disable-next-line react/no-unused-prop-types
  values: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func,
};

Checkboxes.defaultProps = {
  label: '',
  value: [],
  values: [],
  onChange () {},
};

export default Checkboxes;
