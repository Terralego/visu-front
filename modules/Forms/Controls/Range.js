import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { RangeSlider } from '@blueprintjs/core';
import ContentEditable from 'react-contenteditable';

import './index.scss';

const DEFAULT_MIN = 0;
const DEFAULT_MAX = 100;

const handleManualChange = ({
  onChange, prevMin, prevMax, newMin, newMax, min, max,
}) => {
  if (newMin !== undefined) {
    onChange([Math.min(max, Math.min(newMin, prevMax)), prevMax]);
  }
  if (newMax !== undefined) {
    onChange([prevMin, Math.max(min, Math.max(prevMin, newMax))]);
  }
};

const selectAll = ({ target }) => {
  const selection = global.getSelection();
  const range = document.createRange();
  range.selectNodeContents(target);
  selection.removeAllRanges();
  selection.addRange(range);
};

export const Range = ({
  label,
  value,
  onChange,
  min,
  max,
  ...props
}) => {
  const [minValue, maxValue] = value || [min, max];

  if (value && (value.length !== 2 || (+minValue > +maxValue))) {
    throw new Error('Range control: There must be two values and the first must be less than the second');
  }

  const [newMin, setMin] = useState();
  const [newMax, setMax] = useState();
  const [blured, setBlured] = useState(false);

  useEffect(() => {
    if (blured) {
      handleManualChange({
        onChange,
        prevMin: minValue,
        prevMax: maxValue,
        newMin,
        newMax,
        min,
        max,
      });
      setMin();
      setMax();
      setBlured(false);
    }
  }, [blured]);

  return (
    <div className="control-container control--range range">
      <p className="control-label">{label}</p>
      <RangeSlider
        value={value || [min, max]}
        onChange={onChange}
        min={min}
        max={max}
        labelRenderer={itemValue => {
          if (+itemValue === minValue || +itemValue === maxValue) {
            const minChanged = +itemValue === minValue;
            const maxChanged = !minChanged;
            const inputValue = minChanged
              ? newMin
              : newMax;
            return (
              <ContentEditable
                className="range__manual"
                onChange={({ target: { value: newValue } }) => {
                  const fixedNewValue = Number.isNaN(+newValue)
                    ? 0
                    : +newValue;
                  if (minChanged) {
                    setMin(fixedNewValue);
                  }
                  if (maxChanged) {
                    setMax(fixedNewValue);
                  }
                }}
                html={`${inputValue === undefined ? itemValue : inputValue}`}
                onBlur={() => setBlured(true)}
                onFocus={selectAll}
              />
            );
          }
          return itemValue;
        }}
        {...props}
      />
    </div>
  );
};

Range.propTypes = {
  label: PropTypes.string,
  value: PropTypes.arrayOf(PropTypes.number),
  onChange: PropTypes.func,
  min: PropTypes.number,
  max: PropTypes.number,
};

Range.defaultProps = {
  label: '',
  value: null,
  onChange () {},
  min: DEFAULT_MIN,
  max: DEFAULT_MAX,
};

export default Range;
