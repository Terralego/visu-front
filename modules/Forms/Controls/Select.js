import React from 'react';
import { MenuItem, Button } from '@blueprintjs/core';
import { Select } from '@blueprintjs/select';

export class WrapperSelect extends React.Component {
  constructor (props) {
    super(props);
    const { values } = this.props;
    this.state = {
      value: values[0],
    };
  }

  handleChange = value => {
    const { onChange } = this.props;
    this.setState({ value });
    onChange(value);
  }

  handleClick = value => () => this.handleChange(value);

  render () {
    const { label, values = [] } = this.props;
    const { value } = this.state;
    const { handleChange, handleClick } = this;

    return (
      <div>
        <p>{label}</p>
        <Select
          popoverProps={{ usePortal: false }}
          items={values}
          filterable={values.length > 9}
          itemRenderer={val => <MenuItem onClick={handleClick(val)} key={val} text={val} />}
          noResults={<MenuItem disabled text="No results." />}
          onItemSelect={handleChange}
        >
          <Button text={value} rightIcon="double-caret-vertical" />
        </Select>
      </div>
    );
  }
}

export default WrapperSelect;
