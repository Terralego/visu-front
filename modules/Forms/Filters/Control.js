import React from 'react';
import moize from 'moize';

import { TYPE_MANY } from './Filters';

const getValuesWithEmptyItem = moize((values, emptySelectItem) => [{
  value: '',
  label: emptySelectItem,
}, ...values]);

export class Control extends React.Component {
  static getDerivedStateFromProps ({ values, locales: { emptySelectItem }, type }) {
    if (!values) return null;

    const withEmptyValue = values && type !== TYPE_MANY;

    return {
      values: withEmptyValue
        ? getValuesWithEmptyItem(values, emptySelectItem)
        : values,
    };
  }

  state = {
    values: [],
  };

  render () {
    const { component: Component, ...props } = this.props;
    const { values } = this.state;

    return (
      <Component
        {...props}
        values={values}
      />
    );
  }
}

export default Control;
