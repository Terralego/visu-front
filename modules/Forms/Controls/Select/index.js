import Select from './Select';
import './styles.scss';

export default Select;
