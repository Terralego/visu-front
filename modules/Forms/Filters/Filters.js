import React from 'react';
import PropTypes from 'prop-types';

import Select from '../Controls/Select';
import Text from '../Controls/Text';
import Checkboxes from '../Controls/Checkboxes';

export const TYPE_SINGLE = 'single';
export const TYPE_MANY = 'many';
export const TYPE_RANGE = 'range';

const DEFAULT_LOCALES = {
  noResults: 'No results',
};

function getComponent (type, values) {
  switch (type) {
    case TYPE_SINGLE:
      return Array.isArray(values)
        ? Select
        : Text;
    case TYPE_MANY:
      return Checkboxes;
    default:
      return null;
  }
}

export class Filters extends React.Component {
  static propTypes = {
    locales: PropTypes.shape({
      noResults: PropTypes.string,
    }),
    onChange: PropTypes.func,
    layer: PropTypes.string,
    properties: PropTypes.arrayOf(PropTypes.shape({
      property: PropTypes.string.isRequired,
      label: PropTypes.string,
      placeholder: PropTypes.string,

      type: PropTypes.oneOf([TYPE_SINGLE, TYPE_MANY, TYPE_RANGE]).isRequired,
      values: PropTypes.arrayOf(PropTypes.string),
    })),
  }

  static defaultProps = {
    locales: {},
    onChange () {},
    properties: [],
    layer: '',
  }

  state = {
    filters: [],
    properties: {},
  }

  componentDidMount () {
    this.generateFilters();
  }

  onChange = property => value => {
    const { onChange } = this.props;
    this.setState(({ properties: prevProperties }) => {
      const properties = {
        ...prevProperties,
        [property]: value,
      };
      onChange(properties);
      return { properties };
    });
  }

  generateFilters () {
    const { properties: propertiesSchema, locales: customLocales } = this.props;
    const locales = { ...DEFAULT_LOCALES, ...customLocales };
    const filters = propertiesSchema.map(({ type, values, property, ...props }) => {
      const Component = getComponent(type, values);
      const { properties } = this.state;

      if (!Component) return null;


      return (
        <Component
          key={property}
          type={type}
          values={values}
          property={property}
          onChange={this.onChange(property)}
          value={properties[property] || ''}
          locales={locales}
          {...props}
        />
      );
    });
    this.setState({ filters });
  }

  render () {
    const { filters } = this.state;

    return (
      <div>
        {filters}
      </div>
    );
  }
}

export default Filters;
