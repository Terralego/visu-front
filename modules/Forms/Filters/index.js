import Filters from './Filters';

export { TYPE_SINGLE, TYPE_MANY, TYPE_RANGE, TYPE_BOOL } from './Filters';
export default Filters;
