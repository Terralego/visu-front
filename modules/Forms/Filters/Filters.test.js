import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Select from '../Controls/Select';
import Text from '../Controls/Text';
import Checkboxes from '../Controls/Checkboxes';
import Range from '../Controls/Range';
import Switch from '../Controls/Switch';

import Filters, { TYPE_SINGLE, TYPE_MANY, TYPE_RANGE, TYPE_BOOL, getComponent } from './Filters';

jest.mock('@blueprintjs/core', () => ({
  Checkbox: () => <p>Checkbox</p>,
  InputGroup: () => <p>Input</p>,
  Select: ({ children }) => children,
  Switch: () => <p>Switch</p>,
  MenuItem: () => <p>Menu Item</p>,
  Button: () => <p>Button</p>,
  RangeSlider: () => <p>Range</p>,
}));

jest.mock('../Controls/Text', () => () => <p>ControlText</p>);
jest.mock('../Controls/Select', () => () => <p>ControlSelect</p>);
jest.mock('../Controls/Checkboxes', () => () => <p>ControlCheckboxes</p>);
jest.mock('../Controls/Range', () => () => <p>ControlRange</p>);

it('should build a form', () => {
  const tree = renderer.create((
    <Filters
      locales={{ noResults: 'Aucun résultats' }}
      onChange={() => null}
      properties={[{
        property: 'single_value',
        label: 'Vocations',
        type: TYPE_SINGLE,
      }, {
        property: 'single_value_forced',
        label: 'Vocations',
        type: TYPE_SINGLE,
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'many_values',
        label: 'Foo',
        type: TYPE_MANY,
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'range_values',
        label: 'Percentage of bar',
        type: TYPE_RANGE,
        values: [40, 60],
        min: 0,
        max: 100,
      }, {
        property: 'one_value',
        label: 'Booléen',
        type: TYPE_BOOL,
        value: true,
      }]}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should mount & update correctly', () => {
  const wrapper = shallow((
    <Filters
      onChange={() => null}
      properties={[{
        property: 'single_value',
        label: 'Vocations',
        type: TYPE_SINGLE,
      }, {
        property: 'single_value_forced',
        label: 'Vocations',
        type: TYPE_SINGLE,
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'many_values',
        label: 'Foo',
        type: TYPE_MANY,
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'range_values',
        label: 'Percentage of bar',
        type: TYPE_RANGE,
        values: [40, 60],
        min: 0,
        max: 100,
      }, {
        property: 'switch_value',
        label: 'Switch label',
        type: TYPE_BOOL,
      }]}
    />
  ));

  wrapper.instance().onChange('pwout')('pwet');
  expect(wrapper.instance().state.properties.pwout).toEqual('pwet');
});

it('should get component', () => {
  expect(getComponent(TYPE_SINGLE, 'text') === Text).toBe(true);
  expect(getComponent(TYPE_SINGLE, ['text', 'text2']) === Select).toBe(true);
  expect(getComponent(TYPE_MANY, ['text', 'text2']) === Checkboxes).toBe(true);
  expect(getComponent(TYPE_BOOL, true) === Switch).toBe(true);
  expect(getComponent(TYPE_RANGE, [10, 90]) === Range).toBe(true);
  expect(getComponent('NONE_EXISTING_TYPE', 'text') === null).toBe(true);
});
