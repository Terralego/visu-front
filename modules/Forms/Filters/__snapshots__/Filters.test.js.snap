// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build a form 1`] = `
<div>
  <p>
    ControlText
  </p>
  <p>
    ControlSelect
  </p>
  <p>
    ControlSelect with initialContent set to
    <p>
      Enter a query first
    </p>
  </p>
  <p>
    ControlCheckboxes
  </p>
  <p>
    ControlMultiSelect
  </p>
  <p>
    ControlRange
  </p>
  <div
    className="control-container"
  >
    <p>
      Switch
    </p>
  </div>
</div>
`;
